Shifter additions

CodyKank opened the pull request:
# Basic Shifter Execution Support
Changing `files/job_wrapper.sh` to abstract which container we are using for the RJC execution backend. Rather than relying on Singularity being present at all HPC sites Shifter support is added.

## Container Execution
Each container has different execution steps / arguments. Those are handled and passed through the functions `setup_singularity` and `setup_shifter`. To allow these changes in execution, `Singularity` support is no longer hard coded into `reana_job_controller/htcondor_job_manager.py`, and instead of passing raw `Singularity` specific commands only the necessary components are passed as arguments to `files/job_wrapper.sh`. See below:
```
sub['arguments'] = "{0} {1} {2}".format(self.workflow_workspace,self.docker_img, re.sub(r'"', '\\"', self.cmd))
```

## Shifter Issues
As volume mounting / binding in `Shifter` doesn't allow symlinks nor binding to protected directories (`/var` mainly here for `REANA`), the workflow directory mounting will need to be reworked from `/var/reana/. . .` to `/reana/. . .`. More testing is needed for this portion through `bosco`  and other remote submission methods.

CodyKank committed:
Add execution steps for shifter

job_wrapper slightly reworked to abstract the
container being executed. htcondor_job_manager also
modified to not assume singularity execution.

CodyKank committed:
Change image pull method for shifter

Instead of checking for the image with shifterimg lookup
a pull is automatically done, this solves issues of apparent
database misalignments.

CodyKank committed:
Remove debugging messages

CodyKank committed:
Rework directory binding and chdir for containers

To get around shifter v 16.08 not having "--workdir"
available, added support for the ugly
'bash -c "cd $REANA_WD; $CMDS"' for both singularity
and shifter to allow one line for both submissions that
enable each container to chdir directly to the $RWD

CodyKank committed:
Modify container search methods

VC3 now handles certain cases, took out searching
for modules by default. Will now only search if
a container is not found within path (VC3 will do this).

CodyKank committed:
Cleanup comments. Move check for modules

khurtado commented on files/job_wrapper.sh:
```diff
@@ -41,76 +50,128 @@ find_module(){
 tmpjob=$(mktemp -p .)
 chmod +x $tmpjob 
-echo "$container_path" "$@" > $tmpjob
+echo "$CONTAINER_PATH" "$cntr_arguments" "bash -c \"cd /reana; ${@:3}\"" > $tmpjob
```
I think reana already does `bash -c \"cd /reana; ` when setting `self.cmd`. You should be able to see that doing: condor_q -af Args (Arguments?) on submitted jobs.

CodyKank committed:
Reformated singularity and Shifter executions:job_wrapper.sh

Rework execution of containers to properly
execute on different platforms. Changed mounting
of directories for shifter and removed extra
"bash -c" from cmd-list in job_wrapper.

CodyKank committed:
Comment cleanup
